src/workerd/api/sockets.c++

```diff
@@ -0,0 +1,229 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+#include "sockets.h"
+#include "system-streams.h"
+
+
+namespace workerd::api {
+
+
+class TCPConnectResponseImpl final: public kj::HttpService::ConnectResponse, public kj::Refcounted {
+public:
+  TCPConnectResponseImpl(kj::Own<kj::PromiseFulfiller<kj::HttpClient::ConnectResponse>> fulfiller)
+      : fulfiller(kj::mv(fulfiller)) {}
+
+  void setPromise(kj::Promise<void> promise) {
+    task = promise.eagerlyEvaluate([this](kj::Exception&& exception) {
+      if (fulfiller->isWaiting()) {
+        fulfiller->reject(kj::mv(exception));
+      } else {
+        kj::throwRecoverableException(kj::mv(exception));
+      }
+    });
+  }
+
+  kj::Own<kj::AsyncIoStream> accept(
+      uint statusCode, kj::StringPtr statusText, const kj::HttpHeaders& headers) override {
+    KJ_REQUIRE(statusCode >= 200 && statusCode < 300, "the statusCode must be 2xx for accept");
+
+    auto headersCopy = kj::heap(headers.clone());
+
+    auto pipe = kj::newTwoWayPipe();
+
+    fulfiller->fulfill(kj::HttpClient::ConnectResponse {
+      statusCode,
+      statusText,
+      headersCopy.get(),
+      pipe.ends[0].attach(kj::addRef(*this)),
+    });
+    return kj::mv(pipe.ends[1]);
+  }
+
+  kj::Own<kj::AsyncOutputStream> reject(
+      uint statusCode, kj::StringPtr statusText, const kj::HttpHeaders& headers,
+      kj::Maybe<uint64_t> expectedBodySize) override {
+    KJ_REQUIRE(statusCode < 200 || statusCode >= 300, "the statusCode must not be 2xx for reject.");
+    kj::throwRecoverableException(KJ_EXCEPTION(FAILED,
+        kj::str("jsg.Error: Connection rejected by proxy with HTTP code ", statusCode)));
+    KJ_UNREACHABLE;
+  }
+
+private:
+  kj::Own<kj::PromiseFulfiller<kj::HttpClient::ConnectResponse>> fulfiller;
+  kj::Promise<void> task = nullptr;
+};
+
+jsg::Ref<Socket> connectImplNoOutputLock(
+    jsg::Lock& js, jsg::Ref<Fetcher> fetcher, kj::String address) {
+  auto& ioContext = IoContext::current();
+
+  auto jsRequest = Request::constructor(js, kj::str(address), nullptr);
+  kj::Own<WorkerInterface> client = fetcher->getClient(
+      ioContext, jsRequest->serializeCfBlobJson(js), "connect"_kj);
+
+  // TODO: Validate `address` is well formed. Do I need to use `parseAddress` here or is there
```

Comment:
KJ's `parseAddress` probably isn't even what you want here, you should write a validator specific to what this API expects. E.g. we wouldn't want to accept unix socket addresses here.